Atomic record fields

clef-men opened the pull request:
Example:

```ocaml
type t = {
  mutable current : int;
  mutable readers: int [@atomic];
}

let add_new_reader v =
  Atomic.Loc.incr [%atomic.loc v.readers]
```

This PR is implemented by myself (@clef-men) with help from @gasche, including for this PR description.

It is ready for review. We did our best to have a clean git history, so reading the PR commit-by-commit is recommended.

This PR sits on top of #13396, #13397 and #13398.
(Helping make decisions on those PRs will help the present PR move forward.)

(cc @OlivierNicole, @polytypic)

## Why

Current OCaml 5 only supports atomic operations on a special type `'a Atomic.t` of atomic references, which are like `'a ref` except that access operations are atomic. When implementing concurrent data structures, it would be desirable for performance to have records with atomic fields, instead of having an indirection on each atomic field -- just like `mutable x : foo` is more efficient than `x : foo ref`. This is also helpful when combined with inline records inside variant construcors.

## Design

This PR implements the "Atomic Record Fields" RFC
  https://github.com/ocaml/RFCs/pull/39
more specifically the "design number 2" from
  https://github.com/ocaml/RFCs/pull/39#issuecomment-2165173583
proposed by @bclement-ocp.

(The description below is self-contained, reading the RFC and is discussion is not necessary.)

We implement two features in sequence, as described in the RFC.

First, atomic record fields are just record fields marked with an `[@atomic]` attribute. Reads and writes to these fields are compiled to atomic operations. In our example, the field `readers` is marked atomic, to `v.readers` and `v.readers <- 42` will be compiled to an atomic read and an atomic write.

Second, we implement "atomic locations", which is a compiler-supported way to describe an atomic field within a record to perform other atomic operations than read and write. Continuing this example, `[%atomic.loc v.readers]` has type `int Atomic.Loc.t`, which indicates an atomic location of type `int`. The submodule `Atomic.Loc` exposes operations similar to `Atomic`, but on this new type `Atomic.Loc.t`.

Currently the only atomic locations supported are atomic record fields. In the future we hope to expose atomic arrays using a similar approach, but we limit the scope of the current PR to record fields.


## Implementation (high-level)

- In trunk, `Atomic.get` is implemented by a direct memory access, but all other `Atomic` primitives are implemented in C, for example:
  
   `value caml_atomic_exchange(value ref, value newval)`
  
   We preserve this design, but introduce new C functions that take a pointer and an offset instead of just an atomic reference, for example:
   
   `value caml_atomic_exchange_field(value obj, value vfield, value newval)`

   (The old functions are kept around for backward-compatibility reasons, redefined from the new ones with offset `0`.)

- Internally, a value of type `'a Atomic.Loc.t` is a pair of a block and an offset inside the block. With the example above, `[%atomic.loc v.readers]` is the pair `(v, 1)`, indicating the second field of the record `v`. The call `Atomic.Loc.exchange [%atomic.loc v.readers] x` gets rewritten to something like `%atomic_exchange_field v 1 x`, which will eventually become the C call `caml_atomic_exchange_field(v, Val_long(1), x)`. (When an atomic primitive is directly applied to an `[%atomic.loc ...]` expression, the compiler eliminates the pair construction on the fly. If it is passed around as a first-class location, then the pair may be constructed.)

- We reimplement the `Atomic.t` type as a record with a single atomic field, and the corresponding functions become calls to the `Atomic.Loc.t` primitives, with offset `0`.

After this PR, the entire code of `stdlib/atomic.ml` is as follows. (`'a atomic_loc` is a new builtin/predef type, used to typecheck the `[%atomic.loc ..]` construction.)

```ocaml
external ignore : 'a -> unit = "%ignore"

module Loc = struct
  type 'a t = 'a atomic_loc

  external get : 'a t -> 'a = "%atomic_load_loc"
  external exchange : 'a t -> 'a -> 'a = "%atomic_exchange_loc"
  external compare_and_set : 'a t -> 'a -> 'a -> bool = "%atomic_cas_loc"
  external fetch_and_add : int t -> int -> int = "%atomic_fetch_add_loc"

  let set t v =
    ignore (exchange t v)
  let incr t =
    ignore (fetch_and_add t 1)
  let decr t =
    ignore (fetch_and_add t (-1))
end

type !'a t = { mutable contents: 'a [@atomic]; }

let make v = { contents= v }

external make_contended : 'a -> 'a t = "caml_atomic_make_contended"

let get t =
  t.contents
let set t v =
  t.contents <- v

let exchange t v =
  Loc.exchange [%atomic.loc t.contents] v
let compare_and_set t old new_ =
  Loc.compare_and_set [%atomic.loc t.contents] old new_
let fetch_and_add t incr =
  Loc.fetch_and_add [%atomic.loc t.contents] incr
let incr t =
  Loc.incr [%atomic.loc t.contents]
let decr t =
  Loc.decr [%atomic.loc t.contents]
```

There is currently no support for something similar to `Atomic.make_contented` (placing values on isolated cache lines to avoid false sharing) for records with atomic fields . Workflows that require `make_contended` must stick to the existing `Atomic.t` type. Allocation directives for records or record fields could be future work -- outside the scope of the present PR.

gasche committed:
[minor] printcmm: print atomic loads

gasche committed:
[testsuite] test the compilation of atomic functions

clef-men committed:
[refactor] Define [caml_atomic_cas] in terms of [caml_atomic_cas_field].

gasche committed:
[breaking] Change the prototype of [caml_atomic_cas_field].

This is a breaking change because this function was (unfortunately)
exposed outside CAML_INTERNALS, and is used by exactly one external
user, you guessed it:
  https://github.com/ocaml-multicore/multicore-magic/blob/360c2e829c9addeca9ccaee1c71f4ad36bb14a79/src/Multicore_magic.mli#L181-L185
  https://github.com/ocaml-multicore/multicore-magic/blob/360c2e829c9addeca9ccaee1c71f4ad36bb14a79/src/unboxed5/multicore_magic_atomic_array.ml#L36-L43

We chose to change the prototype to remain consistent with the naming
convention for the new caml_atomic_*_field primitives, which will be
added to support atomic record fields.

User code can easily adapt to this new prototype we are using, but not
in a way that is compatible with both old and new versions of
OCaml (not without some preprocessing at least).

Another option would be to expose

    int caml_atomic_cas_field(value obj, intnat fld, value, value)
    value caml_atomic_cas_field_boxed(value obj, value vfld, value, value)

but no other group of primitives in the runtime uses this _boxed
terminology, they instead use

    int caml_atomic_cas_field_unboxed(value obj, intnat fld, value, value)
    value caml_atomic_cas_field(value obj, value vfld, value, value)

and this would again break compatiblity -- it is not easier to convert
code to that two-version proposal, and not noticeably more efficient.

So in this case we decided to break compatibility (of an obscure,
experimental, undocumented but exposed feature) in favor of
consistency and simplificity of the result.

clef-men committed:
runtime: Add atomic primitives [caml_atomic_load_field], [caml_atomic_exchange_field] and [caml_atomic_fetch_add_field].

gasche committed:
lambda: Add support for new atomic primitives.

Uses of existing atomic primitives %atomic_foo, which act on
single-field references, are now translated into %atomic_foo_field,
which act on a pointer and an offset -- passed as separate arguments.

In particular, note that the arity of the internal Lambda primitive
    Patomic_load
increases by one with this patchset. (Initially we renamed it into
    Patomic_load_field
but this creates a lot of churn for no clear benefits.)

We also support primitives of the form %atomic_foo_loc, which
expects a pair of a pointer and an offset (as a single argument),
as we proposed in the RFC on atomic fields
  https://github.com/ocaml/RFCs/pull/39
(but there is no language-level support for atomic record fields yet)

Co-authored-by: Clément Allain <clef-men@orange.fr>

gasche committed:
Add tests for atomic record fields.

gasche committed:
[minor] Make test_untypeast.ml more readable.

clef-men committed:
Add module [Stdlib.Atomic.Loc].

gasche committed:
Add tests for [Atomic.Loc].

clef-men committed:
Reimplement [Atomic.t] using [Atomic.Loc.t].

bclement-ocp commented on typing/typecore.ml:
```diff
@@ -4880,6 +4904,14 @@ and type_label_access env srecord usage lid =
       (Label.disambiguate usage lid env expected_type) labels in
   (record, label, expected_type)
 
+and solve_Pexp_field ~label_usage env sexp srecord lid =
```
Why `solve` rather than `type`?

gasche replied:
`type_<foo>` functions return a typedtree, while `solve_<foo>` functions do the heavy lifting of the type-checking but without constructing a typedtree. (See other `solve_*` functions.) This is useful to factorize code when there are several ways to elaborate the typing rules of a given syntactic construct.

bclement-ocp replied:
Ah, this is just me not being familiar with this file. Thanks!

bclement-ocp commented on typing/types.mli:
```diff
@@ -700,6 +701,7 @@ type label_description =
     lbl_res: type_expr;                 (* Type of the result *)
     lbl_arg: type_expr;                 (* Type of the argument *)
     lbl_mut: mutable_flag;              (* Is this a mutable field? *)
+    lbl_atomic: bool;                   (* Is this an atomic field? *)
```
The use of `bool` in an API where all other flags have their own type (`mutable_flag`, `virtual_flag`, `override_flag`, `closed_flag`, `rec_flag`, `direction_flag`) stands out.

Defining

```ocaml
type atomic_flag = Atomic | Nonatomic
```

and using it throughout would feel more consistent.

bclement-ocp replied:
Given that the `[@atomic]` annotation can only go on mutable fields, I also wonder if it would make sense to have `Atomic` be a third value for the `mutable_flag` type instead of a separate flag. I am not sure it would actually be worth it due to the induced churn, but I am curious if this has been considered.

gasche replied:
My guess is that adding `Atomic` as a third value would make the code a mess, because it entangles two aspect that are more orthogonal than that. (It is not *useful* to have immutable atomics, and we cannot express them without adding more type-level information, but it is certainly something that makes sense.)

bclement-ocp replied:
I think that's fair.

lthls replied:
> (It is not _useful_ to have immutable atomics, and we cannot express them without adding more type-level information, but it is certainly something that makes sense.)

I don't think it makes sense to have immutable atomics at all. Immutable values have semantics that are completely incompatible with atomic behaviour. I'm happy with keeping two flags though: it's a bit of a waste of memory but there are lots of cases where mutable values are treated the same, atomic or not.

bclement-ocp commented on stdlib/atomic.ml:
```diff
@@ -12,16 +12,45 @@
+end
+
+type !'a t =
+  { mutable contents: 'a [@atomic];
```
The parallel with `ref` feels somewhat satisfying!

bclement-ocp commented on ocamldoc/odoc_sig.ml:
```diff
@@ -433,7 +433,13 @@ module Analyser =
+         ld_atomic=is_atomic;
+         ld_type=type_expr;
+         ld_attributes} =
+      assert (not is_atomic);
```
I don't really understand why this `assert` would be true, can you explain?

gasche replied:
Good catch, this was intended as a TODO marker, we have to complete this case with proper support for atomic fields below (add an `atomic` field in `Odoc_type.record_field`, and then handle it in the consumers somehow).

kayceesrk commented:
Thanks for this contribution. I am starting to review this PR.

I wondered why `[@atomic]` attribute was necessary at all and why we can't just use `[%atomic.loc ...]` on arbitrary fields to perform the atomic operations. One answer is that the OCaml memory model separates atomic and non-atomic locations. If we allow `[%atomic.loc ...]` on arbitrary fields, then we open up the possibility of mixed-mode accesses, which the memory model doesn't specify the semantics of. 

I also wondered whether it was considered to introduce `atomic` as a keyword to annotate atomic record fields:

```
type t = {
  mutable current : int;
  atomic readers: int;
}
```

instead of 

```
type t = {
  mutable current : int;
  mutable readers: int [@atomic];
}
```
I'm not suggesting that we do this, and only curious to understand the choice.

kayceesrk commented on testsuite/tests/atomic-locs/record_fields.ml:
```diff
@@ -0,0 +1,195 @@
+end
+[%%expect{|
+Line 2, characters 13-30:
+2 |   type t = { x : int [@atomic] }
```
Would be good to have more test coverage for the following cases:

```
 % rlwrap ../_install/bin/ocaml
OCaml version 5.4.0+dev0-2024-08-25
Enter #help;; for help.

# type t = {
    mutable current : int;
    mutable readers : int [@atomic];
  };;
type t = { mutable current : int; mutable readers : int [@atomic]; }
# let f v = Atomic.Loc.get [%atomic.loc v.current];;
Error: The record field current is not atomic
# let f v = Atomic.Loc.get [%atomic.loc v];;
Error: Invalid [%atomic.loc] payload, a record field access is expected.
```

bclement-ocp reviewed (commented):
@gasche suggested I take a look at this since I suggested this variant of the design. The implementation looks like it faithfully implements the described changes; I don't have many comments on the code that seems fairly straightforward.

With this patch, all regular atomic writes now incur an additional integer untagging operation. This is briefly discussed and brushed away in one of the commits as being "not noticeably more efficient". It is probably true for most (all?) hardware and workflows, especially given that this is inside a C call, but can you share the measurements (if any) that were made to support that assessment? I would personally err on the side of caution here and provide untagged variants, especially given that atomics are likely to be used for performance.

I noted while reading the PR that people who need the extra flexibility of proposal 1 (first-class location) from https://github.com/ocaml/RFCs/pull/39 can do:

```ocaml
module Atomic_field = struct
  type ('r, 'a) t

  external get : 'r -> ('r, 'a) t -> 'a = "%atomic_load_field"
  external exchange : 'r -> ('r, 'a) t -> 'a -> 'a = "%atomic_exchange_field"
  external compare_and_set : 'r -> ('r, 'a) t -> 'a -> 'a -> bool = "%atomic_cas_field"
  external fetch_and_add : 'r -> ('r, int) t -> int -> int = "%atomic_fetch_add_field"
end
```

The only missing piece would be an `[%atomic.field bar]` extension point, although I don't think it necessarily makes sense to add it in this PR (especially given that there are some interactions e.g. with type disambiguation that would need to be worked out first). It was not necessarily obvious to me initially that both designs could work together this way.

gasche commented:
> With this patch, all regular atomic writes now incur an additional integer untagging operation. [..]  It is probably true for most (all?) hardware and workflows, especially given that this is inside a C call, but can you share the measurements (if any) that were made to support that assessment?

No measurements, I assess that the cost of the OCaml/C context switch and of the atomic operation itself dwarf the cost of a bit shift. (I think that if people wanted to make the code more complex because they believe otherwise, they should try to disprove this with microbenchmarks first.)

clef-men commented:
> Thanks for this contribution. I am starting to review this PR.
> 
> I wondered why `[@atomic]` attribute was necessary at all and why we can't just use `[%atomic.loc ...]` on arbitrary fields to perform the atomic operations. One answer is that the OCaml memory model separates atomic and non-atomic locations. If we allow `[%atomic.loc ...]` on arbitrary fields, then we open up the possibility of mixed-mode accesses, which the memory model doesn't specify the semantics of.
> 
> I also wondered whether it was considered to introduce `atomic` as a keyword to annotate atomic record fields:
> 
> ```
> type t = {
>   mutable current : int;
>   atomic readers: int;
> }
> ```
> 
> instead of
> 
> ```
> type t = {
>   mutable current : int;
>   mutable readers: int [@atomic];
> }
> ```
> 
> I'm not suggesting that we do this, and only curious to understand the choice.

I think I would be more convenient to introduce the `atomic` keyword.
We considered it but were concerned that it would be difficult to have it accepted. We chose the easy, non-invasive way.
However, if everyone agrees on it, we can include it in the PR.

clef-men commented:
> Thanks for this contribution. I am starting to review this PR.
> 
> I wondered why `[@atomic]` attribute was necessary at all and why we can't just use `[%atomic.loc ...]` on arbitrary fields to perform the atomic operations. One answer is that the OCaml memory model separates atomic and non-atomic locations. If we allow `[%atomic.loc ...]` on arbitrary fields, then we open up the possibility of mixed-mode accesses, which the memory model doesn't specify the semantics of.
> 
> I also wondered whether it was considered to introduce `atomic` as a keyword to annotate atomic record fields:
> 
> ```
> type t = {
>   mutable current : int;
>   atomic readers: int;
> }
> ```
> 
> instead of
> 
> ```
> type t = {
>   mutable current : int;
>   mutable readers: int [@atomic];
> }
> ```
> 
> I'm not suggesting that we do this, and only curious to understand the choice.

I think I would be more convenient to introduce the `atomic` keyword.
We considered it but were concerned that it would be difficult to have it accepted. We chose the easy, non-invasive way.
However, if everyone agrees on it, we can include it in the PR.

fpottier commented:
In principle I would be in favor of making `atomic` a keyword. I don't think an attribute should influence the semantics of the code.

I would even go so far as to suggest that one should write `atomic mutable` or `mutable atomic` as opposed to just `atomic`.

I would also be happy if we could find a better syntax for `[%atomic.loc x.foo]`. I am not sure what to propose. By analogy with C, perhaps `&x.foo` would make sense, but I suppose that this is impossible, for compatibility reasons?

gasche commented:
My appetite for syntax dicussions is fairly low. Maybe we could first consider the feature and its implementation, merge the present PR if we decide to go for it, and *then* discuss reserved syntax for those constructs? If a release were to happen in between the two changes, maintaining both options is very low-maintenance so it's no big deal.

(Another option would be to move syntax discussions to https://github.com/ocaml/RFCs/pull/39, in which case they could start right away without flooding the discussion of this one implementation.)

fpottier commented:
Sure, my comments about syntax were not meant to delay this PR.

I just happen to think that syntax *is* an important aspect of language design, so I hope that (sooner rather than later) it is possible to give a nice syntactic appearance to the proposed new features.

clef-men committed:
atomic fields in ocamldoc

OlivierNicole commented on runtime/memory.c:
```diff
@@ -314,36 +314,43 @@ CAMLexport CAMLweakdef void caml_initialize (volatile value *fp, value val)
     atomic_thread_fence(memory_order_release); /* generates `dmb ish` on Arm64*/
   }
-  write_barrier(ref, 0, ret, v);
+  write_barrier(obj, field, ret, v);
```
It seems ok not to register roots in this function because
1. `write_barrier` does not trigger a GC
2. it is used when all values but `ret` are dead anyway

OlivierNicole commented on lambda/translprim.ml:
```diff
@@ -658,6 +680,75 @@ let lambda_of_loc kind sloc =
+  in
+  match (kind : atomic_kind) with
+  | Record ->
+      (* the primitive application
```
Maybe clarify with
```suggestion
      (* For example, the primitive application
```
here and below

OlivierNicole replied:
Unsure why this was marked as resolved. Do you disagree?

gasche replied:
My memories from our apply-comments-and-rebase session is as follows: (1) yes, we were not fully enthusiastic about the proposal which breaks regularity and reads a bit heavily, but (2) rather than just ignoring it, we decided to rewrite the comments so that they are easier to understand, and in particular @clef-men suggested writing the examples as pseudo-AST rather than pseudo-code, and I think they read nicer, and finally (3) I assumed that the significant changes to the comment had made the point moot, in the sense that the structure would now be different anyway, and we wouldn't have to be fully explicit on (1) (do we really disagree with Olivier here?) when it's maybe not a key part of the PR discussion, so I decided to consider the comment morally "outdated" and make it go away. In retrospect (looking at the old and new comments) the changes are not as invasive as I thought they were, and (3) is not fully convincing, but I propose to keep it as-is and look the other way, if that's fine with you. (Otherwise please do insist and I think we'll come back to add "For example," in each case.)

OlivierNicole replied:
OK, I won’t press it.

OlivierNicole commented on asmcomp/cmm_helpers.mli:
```diff
@@ -214,6 +214,10 @@ val set_field :
 
+(** [field_address_computed ptr ofs dbg] returns an expression for the address
+    at offset [ofs] of the block pointed to by [ptr]. *)
+val field_address_computed : expression -> expression -> Debuginfo.t -> expression
```
Document the fact that the block is assumed not to contain unboxed floats.

lthls replied:
We don't support 32-bit architectures in the native compiler anymore, so it would actually work for flat float records too. But it would be nice to document that the offset is in machine words, and maybe that the resulting expression is a derived pointer (of type `Addr`).

OlivierNicole commented on typing/typedecl.ml:
```diff
@@ -360,7 +367,9 @@ let transl_declaration env sdecl (id, uid) =
-        | [{pld_mutable = Immutable}] -> ()
+        | [{pld_mutable = Immutable; pld_attributes = attrs}] ->
+            if Builtin_attributes.has_atomic attrs then
+              bad "it has an atomic field"
```
This looks like dead code: you raise an error if a field is atomic but not mutable, and the mutable case is handled here, just above.

OlivierNicole commented on typing/typecore.ml:
```diff
@@ -187,6 +187,8 @@ type error =
   | Unrefuted_pattern of pattern
   | Invalid_extension_constructor_payload
   | Not_an_extension_constructor
+  | Invalid_atomic_loc_payload
+  | Label_not_atomic of Longident.t
```
Maybe move this to the relevant, later commit.

OlivierNicole commented on typing/typecore.ml:
```diff
@@ -6966,6 +6968,13 @@ let report_error ~loc env = function
+        Style.inline_code "[%atomic.loc]"
+  | Label_not_atomic lid ->
+      Location.errorf ~loc "The record field %a is not atomic"
+        (Style.as_inline_code longident) lid
```
Also move to the relevant commit

OlivierNicole commented on ocamldoc/odoc_str.ml:
```diff
@@ -168,9 +168,10 @@ let string_of_record l =
+          P.sprintf "   %s%s : %s %s;%s"
             (if field.M.rf_mutable then "mutable " else "") field.M.rf_name
             (Odoc_print.string_of_type_expr field.M.rf_type)
+            (if field.M.rf_atomic then "[@atomic] " else "")
```
This seems to change the printing of records even without atomic fields (because of the space), can you fix that?

OlivierNicole commented on ocamldoc/odoc_html.ml:
```diff
@@ -1707,6 +1707,7 @@ class html =
         if r.rf_mutable then bs b (self#keyword "mutable&nbsp;") ;
         bp b "<span id=\"%s\">%s</span>&nbsp;: " (gen_name r) r.rf_name;
         self#html_of_type_expr b father r.rf_type;
+        if r.rf_atomic then bs b (self#keyword "[@atomic]&nbsp;") ;
```
Check position of non-breaking space wrt the rest

OlivierNicole commented on typing/typecore.ml:
```diff
@@ -4377,6 +4376,29 @@ and type_expect_
+                  )
+               } ] ->
+          let record, label, ty_arg =
+            solve_Pexp_field ~label_usage:Env.Mutation env sexp srecord lid
```
In theory passing `Env.Mutation` here could cause a spurious warning "the record field foo is never mutated" if an atomic record field is only mutated through `Atomic.Loc.set [%atomic.loc x.foo] 42` (rather than `x.foo <- 42`) and the type is not exported. Unlikely but it may happen in generated code, so probably worth at least a TODO.

OlivierNicole replied:
When no change was made, could you please not mark as resolved without an explanation? I’m sure you have reasons to consider this point minor, but then I have to figure out whether my comment was made irrelevant by later changes, or whether you just disagree and decided to silently dismiss my suggestion.

gasche replied:
There was a change made, it shows up two lines below: `Env.mark_label_used Env.Projection label.lbl_uid;`. The Github UI is not ideal, and it would have been easier for you to check things if we had waited before rebasing the review changes into the PR commits. Apologies.

OlivierNicole replied:
I see, I had missed this. With the `Env.Projection` there and the `Env.Mutation` above, am I correct in understanding that there should never be a warning “record field `foo` is never used/mutated” if there is at least one `[%atomic.loc x.foo]` in the program?

gasche replied:
Yes, this is the intent: mark those atomic-loc expressions as performing both read- and write-access to the field. (Because they can perform either, most likely both, and we cannot easily track which.)

OlivierNicole commented on testsuite/tests/atomic-locs/record_fields.ml:
```diff
@@ -11,18 +11,37 @@ module Basic = struct
 
   let set (type a) (r : a atomic) (v : a) : unit = r.x <- v
+
+  let cas (type a ) (r : a atomic) oldv newv =
```
```suggestion
  let cas (type a) (r : a atomic) oldv newv =
```

OlivierNicole commented on testsuite/tests/atomic-locs/record_fields.ml:
```diff
@@ -11,18 +11,37 @@ module Basic = struct
+  let[@inline never] get_loc (type a) (r : a atomic) : a Atomic.Loc.t =
+    [%atomic.loc r.x]
+
+  let slow_cas (type a ) (r : a atomic) oldv newv =
```
idem

OlivierNicole commented on stdlib/atomic.ml:
```diff
@@ -30,15 +30,27 @@ module Loc = struct
+  }
+
+let make v =
+  { contents= v }
```
style (sorry)

OlivierNicole commented on lambda/translcore.ml:
```diff
@@ -330,7 +330,7 @@ and transl_exp0 ~in_new_scope ~scopes e =
       let loc = of_location ~scopes e.exp_loc in
       Lprim (Pmakeblock (0, Immutable, shape), [arg; lbl], loc)
-  | Texp_field (arg, _, lbl) when lbl.lbl_atomic ->
+  | Texp_field (arg, _, lbl) when lbl.lbl_atomic = Atomic ->
```
I would prefer it without polymorphic equality, we could replace `lbl` with `{ lbl_atomic = Atomic; _ }`.

OlivierNicole commented on lambda/translcore.ml:
```diff
@@ -348,7 +348,7 @@ and transl_exp0 ~in_new_scope ~scopes e =
                  of_location ~scopes e.exp_loc)
       end
-  | Texp_setfield (arg, _, lbl, newval) when lbl.lbl_atomic ->
+  | Texp_setfield (arg, _, lbl, newval) when lbl.lbl_atomic = Atomic ->
```
Same remark on `= Atomic`.

OlivierNicole commented on boot/ocamlc:
For some reason, I don’t get the same `boot/ocamlc` and `boot/ocamllex` than in your first bootstrap commit, when I run the boostrap on my machine. The process I followed was: let’s call A the parent commit of the first "Bootstrap" commit, I checked out A’s parent, ran `make world`, then checked out A, ran `make bootstrap`, and `git diff 9470efd432 --` shows differences.

OlivierNicole reviewed (commented):
I reviewed today with @clef-men and @gasche present, and the implementation looks correct to me. I like the design, as I had expressed in the RFC thread.

I have mostly minor comments, apart from the fact that I have been unable to reproduce the first bootstrap commit (see my dedicated review comment about that).

fpottier commented:
Naive question: why is `Atomic.Loc.set` implemented in terms of `Atomic.Loc.exchange`? Isn't that needlessly costly?

OlivierNicole commented:
It reflects the fact that `Atomic.set` is implemented in terms of `Atomic.exchange`, to enforce some properties of the OCaml memory model (see #10995).

gasche committed:
[refactor] more general type for Env.mark_*_used

We want to use [mark_label_used] in a context where we cannot easily
find the label declaration, only the label description (from the
environment).

gasche commented:
This morning I gave a talk on The Pattern-Matching Bug, and we realized (discussion with @clef-men and @fpottier) that there is a connection to atomic record fields:
1. This PR does not handle reads to atomic record fields that come from pattern-matching, only those in `Pexp_getfield` expressions. Oops! At minima one would expect those reads to be atomic reads.
2. It is probably not a good idea to synthesize atomic reads in the middle of pattern-matching control flow. For example if `a` and `b` are atomic fields of the same record, then the pattern `{a; b}` would perform two atomic fields in an implementation-defined order. Instead we propose to forbid all atomic field patterns (even those binding just variables), so users have to bind the record and then access it in the right-hand side.

Forbidding reading atomic fields in patterns is restrictive, we could relax the restriction later if users find it too strong and we understand better what would be reasonable. One unfortunate effect of this restriction is that adding the `[@atomic]` attribute to a record field that already exists may cause the program to be rejected statically, if the code contains pattern-matches on that field.

OlivierNicole commented:
> 2\. It is probably not a good idea to synthesize atomic reads in the middle of pattern-matching control flow. For example if `a` and `b` are atomic fields of the same record, then the pattern `{a; b}` would perform two atomic fields in an implementation-defined order.

Naive question: why is this an issue? If the order of two atomic reads is critical, I would expect a parallel programmer not to perform them via pattern matching.

gasche commented:
I guess it doesn't help that we were discussing The Pattern-Matching Bug at the same time, whose longer name could be What Can Go Horribly Wrong If You Pattern-Match On Mutable Fields. A serious suggestion that emerged during the discussion is to simply forbid all matches on mutable fields in OCaml. I pointed out that this would break too many programs, so that we cannot do this right now, but maybe (the discussion went) it should have been forbidden from the start. In this context, the decision to forbid implicit atomic reads in patterns makes sense, I think -- if we identify a pain point, let's not make the problem worse.

(This being said, I agree with you that this is not the only reasonable choice. In particular, `Atomic.read x + Atomic.read y` also has unspecified atomic-read order and everyone is fine with that.)

gasche commented:
The osx-arm64 fails due to a difference in cmm stamps (I'm not sure why an architecture has different stamps from the other, but oh well). We fixed previous such failures by using predicates to rule out different-cmm-output configurations (we use `not-windows; no-flambda; no-tsan;`. But there is no flag to disable OSX, and no one has been willing to review "negation in ocamltest" (#13315) yet, so there is no easy way to silence this one. I will propose instead to restrict the test to a single known-good configuration (linux-amd64).

(In the long run we should try to change the cmm type definitions to keep more structure to the identifiers, to support the `-dno-unique-ids` option in `-dcmm`, to have more robust test output, or maybe to have an entirely different printer designed with reproducibility in mind.)

lthls commented:
> In particular, `Atomic.read x + Atomic.read y` also has unspecified atomic-read order and everyone is fine with that.

I'm not ! But I concur that I can't get many other people to care about this kind of issues.

clef-men committed:
Forbid atomic fields in patterns.

gasche commented on typing/typecore.ml:
```diff
@@ -986,6 +987,8 @@ let solve_Ppat_construct ~refine tps penv loc constr no_existentials
 
 let solve_Ppat_record_field ~refine loc penv label label_lid record_ty =
   with_local_level_generalize_structure begin fun () ->
+    if not refine && label.lbl_atomic = Atomic then
```
The `not refine` test tells us that we are type-checking a field pattern coming from user code, rather than checking a counter-example candidate for exhaustivity checking. I would propose a comment such as follows to clarify this:

```ocaml
(* [not refine] holds when the pattern comes from the user,
  rather than a counter-example to exhaustivity. *)

gasche commented on typing/typecore.ml:
```diff
@@ -7037,6 +7040,9 @@ let report_error ~loc env = function
         (Style.as_inline_code longident) lid
+  | Atomic_in_pattern lid ->
+      Location.errorf ~loc "Atomic load on field %a in pattern"
+        (Style.as_inline_code longident) lid
```
My proposal for a error wording:

> Atomic fields (here %a) are forbidden in patterns, as it is difficult to reason about when the atomic read will happen during pattern matching: the field may be read zero, one or several times depending on the patterns around it.

gasche commented:
A proposal for a Changes entry:

```
- RFCs#39, #13404: atomic record fields
  (Clément Allain and Gabriel Scherer, review by KC Sivaramakrishnan, Basile Clément
   and Olivier Nicole)
```

gasche committed:
Changes

OlivierNicole commented:
> I guess it doesn't help that we were discussing The Pattern-Matching Bug at the same time, whose longer name could be What Can Go Horribly Wrong If You Pattern-Match On Mutable Fields. A serious suggestion that emerged during the discussion is to simply forbid all matches on mutable fields in OCaml. I pointed out that this would break too many programs, so that we cannot do this right now, but maybe (the discussion went) it should have been forbidden from the start. In this context, the decision to forbid implicit atomic reads in patterns makes sense, I think -- if we identify a pain point, let's not make the problem worse.

In that case, forbidding them in patterns seems reasonable, I guess.

OlivierNicole reviewed (commented):
My previous comments have been addressed (except for a minor point) and the added code to forbid matching on atomic fields looks good to me. There is a small conflict with trunk to resolve.

gasche commented:
Consensus from the maintainer meeting: mild consensus in favor of the feature. People wondered about whether the current proposal (where the record value and the offset are packed together in an `'a Loc.t` type, proposed by @bclement-ocp ) is better than the previous proposal (with only the offset at a `('r, 'a) Loc.t` type), and whether the previous could be better in some scenarios. We are planning to ask @polytypic about this.

(I think that the just-the-offset version is more complex to use for users, and that it makes it harder to extend to arrays.)

gasche commented:
A status report:

- I did implement the alternative `[%atomic.field f]` approach in #13707.

- I realized that it does not currently work with inline records, and proposed a relaxation of their typing rules in #13770. While conceptually simple, this approach required delicate changes in the type-system implementation (#13781). This work is not currently merged, it is in the process of becoming mergeable, but difficulties around inline record fields are not fully solved. Finally, an alternative route regarding inline record (allow to refer explicitly to their type) has been discussed but remains fairly uncertain.

- In view of these design and implementation difficulties, I proposed at our last maintainer meeting (end of March) to go ahead with the current, simpler proposal, implementing `[%atomic.loc r.f]`.

- (If eventually we were able to resolve the inline-record difficulties, we could consider reintroducing `[%atomic.field f]`, with `[%atomic.loc r.f]` expressed as a derived construct on top of it, but whether this would be judged worth it remains to be seen.)

After months of waiting, I am interested to see if we can merge this PR reasonably quickly. I am thinking of coordinating with @clef-men to get this PR rebased on the current trunk and reviewed next week -- ideally before the branching of 5.4.

@OlivierNicole did an excellent review earlier and not much has changed, but (1) Olivier has not approved yet (although https://github.com/ocaml/ocaml/pull/13404#pullrequestreview-2356701018 clearly states his intent to approve) and (2) he will be in holidays next week. I will take care of reviewing the new changes after Olivier's review (notably the disabling of deep subpatterns in atomic fields of record patterns), but ideally we would get a second look from someone who is not a co-author of this PR, with an approval either from their own review or on behalf of Olivier. I wonder if either of @lthls or @nojb would be interested -- I can be available for a synchronous code discussion, in-person or remotely.

gasche commented:
No strong enthusiasm from other potential reviewers. My new plan is to wait for @OlivierNicole to return from holidays and conclude his review, and kindly ask @Octachron if this can be merged after the feature branch -- I tentatively marked this for the 5.4 milestone, no problem with removing if the suggestion does not fly.

clef-men committed:
Add support for atomic record fields.

Requires a bootstrap.

Co-authored-by: Gabriel Scherer <gabriel.scherer@gmail.com>

clef-men committed:
Bootstrap.

To reproduce (see BOOTSTRAP.adoc for details):
- go to the earlier commit "lambda: Add support for new atomic
  primitives" as a known-good state, build that one
- keep the build artifacts around, come here and do `make bootstrap`

Octachron commented:
Ok, for waiting for @OlivierNicole to be back from holidays to confirm he converts his previous review comment to a formal approval.

OlivierNicole commented:
I’m back and happy to look at the few last changes and approve, but I’m having trouble determining what these changes are, as their have been several rebasings since my last review. Don’t hesitate to reach out in private for synchronous discussion.

OlivierNicole commented:
Also, it looks like this is still in need of conflict resolution with trunk before it can be merged.

clef-men committed:
Add predefined type ['a atomic_loc].

This type will be used for ['a Atomic.Loc.t], as proposed
in the RFC
  https://github.com/ocaml/RFCs/pull/39

We implement this here to be able to use it in the stdlib later,
after a bootstrap.

clef-men committed:
Add extension [%atomic.loc record.field].

clef-men committed:
Bootstrap.

This bootstrap is not required by a compiler change, but it enables
the use of the predefined type `'a atomic_loc` and the
expression-former [%atomic.loc ...] in the standard library.

gasche commented:
Regarding your review: I believe that the only feature part that you have not reviewed previously is the last commit, "Forbid atomic fields in patterns.", which came after your review work. The rest corresponds to what you have reviewed, with some cleanups/clarifications following your review comments.

I will try to have a look at the PR myself tonight, and see what should be done to get the CI working again.

OlivierNicole commented:
```
# type t = { x : int; mutable y : int [@atomic] };;
type t = { x : int; mutable y : int [@atomic]; }
# function { x; y = _ } -> x;;
Error: Atomic fields (here y) are forbidden in patterns, as it is difficult to reason about when the atomic read will happen during pattern matching: the field may be read zero, one or several times depending on the patterns around it.
# function { x; _ } -> x;;
```

I find this rather frustrating. Unless I’m missing something, I should be able to explicitly ignore `y`. It is often better to explicitly ignore some fields rather than having a catch-all, when one wants to be notified in case new fields get added.

OlivierNicole commented:
Apart from my previous message, I have no other qualms with the code.

gasche committed:
testsuite for pattern-matching on atomic reads

gasche committed:
allow wildcard patterns under atomic record fields

Suggested-by: Olivier Nicole <olivier@chnik.fr>

gasche commented:
@clef-men, @OlivierNicole: I have
- fixed a test (compiler-libs/test_untypeast.ml) that needed some extra love after the rebase, so maybe the CI passes now (otherwise I'll fix it)
- changed the implementation of the error on atomic fields in patterns to allow for the special case suggested by @OlivierNicole

gasche commented:
(Any non-trivial change goes into a new commit so that it's easy for @OlivierNicole to figure out what to look at.)

OlivierNicole reviewed (approved):
The added code to forbid atomic fields in patterns looks good to me.

Minor: I would add a test for the case `{ x; _ }` as well, for the sake of completeness.

OlivierNicole commented:
The TSan failures are clearly unrelated to this PR.

gasche committed:
stdlib/Atomic: move the Loc module below in the .mli

gasche committed:
stdlib/Atomic: minimal documentation

gasche commented:
I pushed a final change to `stdlib/atomic.mli`: the Loc module is moved to the end of the file, and I added minimal documentation comments. (I tried to include a small but full example, and ran into what appears to be a parsing/lexing bug in ocamldoc, to be reported separately.)

gasche commented:
I propose some documentation of atomic record fields in the reference manual in a separate PR, #13991.

gasche commented:
My current thinking on this PR:

- Now that @OlivierNicole approved the PR, I believe that it is mergeable and within the pre-approved time frame for backporting into the 5.4 release branch. I think it's useful to merge it for 5.4, because it is a useful feature put in the hand of our users -- not just an ergonomic improvement, it changes the way people can write concurrent code.
- But the PR does not include the documentation for the feature, which we didn't write beforehand (our mistake!) and is now a separate PR #13991. I am planning to see if we can get a first round of feebdack early next week on the documentation, to avoid a situation where the feature gets merged but the documentation lingers indefinitely. (Sometimes we don't have documentation because the implementors did not write it, but sometimes it gets stuck in the review pipeline because reviewers demand various changes and authors are not motivated to rework their documetnation.)
- Technically this PR extends the standard library with an `Atomic.Loc` submodule, so it is reasonable to wonder whether this specific part needs approval from a second maintainer. My current thinking is that an approval (from someone else than @OlivierNicole) on the documentation would effectively double as a review of the `Atomic.Loc` signature.

In summary:
- I think we could reasonably click "merge" right now.
- But I'm willing to see if we can get the documentation reviewed before the deadline before going ahead.

OlivierNicole commented:
> * Technically this PR extends the standard library with an `Atomic.Loc` submodule, so it is reasonable to wonder whether this specific part needs approval from a second maintainer. My current thinking is that an approval (from someone else than @OlivierNicole) on the documentation would effectively double as a review of the `Atomic.Loc` signature.

I think either here on or on the documentation PR makes sense (so long as we have a second approval)

gasche commented:
I want to merge this now, but there is now a bootstrap conflict coming from the magic number update which happened *both* on the 5.4 and trunk branches. Grmph...

Octachron commented on stdlib/atomic.mli:
```diff
@@ -68,6 +68,33 @@ val incr : int t -> unit
+      type [foo], then [[%atomic.loc r.f]] has type [foo Atomic.Loc.t].
+
+      The API below mirrors the API of the {!Atomic} module, and
+      requires no further documentation. *)
```
It is more the API on atomic references which is mirrored rather than the whole module, so I would propose

> The API below mirrors the API for {{!t}atomic references} , see the documentation above for more information.

And personally, I think it would be better to mirror the documentation comment on the seven function (with a possible exception for CAS) since the documentation is rather short and it is faster to read it rather than looking for the mirror documentation.

```
  ...
 The API below mirrors the API for {{!t}atomic references} *)
 
(** Get the current value of the atomic location. *)
  external get : 'a t -> 'a = "%atomic_load_loc"
  
  (** Set a new value for the atomic location. *)
  val set : 'a t -> 'a -> unit
  
  (** Set a new value for the atomic location, and return the current value. *)
  external exchange : 'a t -> 'a -> 'a = "%atomic_exchange_loc"
  
  (** [compare_and_set r seen v] sets the new value of [r] to [v] only
    if its current value is physically equal to [seen] -- the
    comparison and the set occur atomically. Returns [true] if the
    comparison succeeded (so the set happened) and [false]
    otherwise. *)
  external compare_and_set : 'a t -> 'a -> 'a -> bool = "%atomic_cas_loc"
  
  (** [fetch_and_add r n] atomically increments the value of [r] by [n],
    and returns the current value (before the increment). *)
  external fetch_and_add : int t -> int -> int = "%atomic_fetch_add_loc"

(** [incr r] atomically increments the value of [r] by [1]. *)
  val incr : int t -> unit
  
  (** [decr r] atomically decrements the value of [r] by [1]. *)
  val decr : int t -> unit
  ```

gasche replied:
I reused the more precise reference. On the other hand, I chose not to copy the mli documentation, because I have a future state in mind where there is a similar `Atomic.Array` submodule that with `get : 'a t -> int -> 'a`, etc., and prefer not to copy the same comments in three different places in total.

As @Octachron suggested (private communication), I reordered the operations to follow exactly the Atomic.t documentation order.

gasche commented:
I have now rebased this PR and replayed the two bootstraps. There is also a branch that was rebased and replayed on top of 5.4: https://github.com/gasche/ocaml/commits/atomic_fields_11-5.4/ . This should be good to merge in trunk and 5.4 if the CI agrees.

gasche commented:
The CI has the "parsetree-updated" check failing. I believe that this is a false negative: @clef-men added `type atomic_flag = Nonatomic | Atomic` to parsing/Asttypes, but this type is only used in the typedtree, there is no parsetree change, so my understanding is that there will be nothing to do in ppxlib.

Octachron commented:
I agree that the parsetree-change failure is a false negative.

gasche commented:
We now have a PR for atomic arrays open at #14380.
